File: app/src/main/java/com/zeapo/pwdstore/pwgen/PasswordGenerator.java
package com.zeapo.pwdstore.pwgen;

import android.content.Context;
import android.content.SharedPreferences;

import java.util.ArrayList;

public class PasswordGenerator {
    static final int DIGITS     = 0x0001;
    static final int UPPERS     = 0x0002;
    static final int SYMBOLS    = 0x0004;
    static final int AMBIGUOUS  = 0x0008;
    static final int NO_VOWELS  = 0x0010;

    static final String DIGITS_STR = "0123456789";
    static final String UPPERS_STR = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    static final String LOWERS_STR = "abcdefghijklmnopqrstuvwxyz";
    static final String SYMBOLS_STR = "!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~";
    static final String AMBIGUOUS_STR = "B8G6I1l0OQDS5Z2";
    static final String VOWELS_STR = "01aeiouyAEIOUY";

    // No a, c, n, h, H, C, 1, N
    private static final String pwOptions = "0ABsvy";

    /**
     * Sets password generation preferences.
     *
     * @param ctx     context from which to retrieve SharedPreferences from
     *                preferences file 'PasswordGenerator'
     * @param argv    options for password generation
     *                <table summary="options for password generation">
     *                <tr><td>Option</td><td>Description</td></tr>
     *                <tr><td>0</td><td>don't include numbers</td></tr>
     *                <tr><td>A</td><td>don't include uppercase letters</td></tr>
     *                <tr><td>B</td><td>don't include ambiguous charactersl</td></tr>
     *                <tr><td>s</td><td>generate completely random passwords</td></tr>
     *                <tr><td>v</td><td>don't include vowels</td></tr>
     *                <tr><td>y</td><td>include at least one symbol</td></tr>
     *                </table>
     * @param numArgv numerical options for password generation: length of
     *                generated passwords followed by number of passwords to
     *                generate
     * @return <code>false</code> if a numerical options is invalid,
     * <code>true</code> otherwise
     */
    public static boolean setPrefs(Context ctx, ArrayList<String> argv
            , int... numArgv) {
        SharedPreferences prefs
                = ctx.getSharedPreferences("PasswordGenerator", Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = prefs.edit();

        for (char option : pwOptions.toCharArray()) {
            if (argv.contains(String.valueOf(option))) {
                editor.putBoolean(String.valueOf(option), true);
                argv.remove(String.valueOf(option));
            } else {
                editor.putBoolean(String.valueOf(option), false);
            }
        }
        for (int i = 0; i < numArgv.length && i < 2; i++) {
            if (numArgv[i] <= 0) {
                // Invalid password length or number of passwords
                return false;
            }
            String name = i == 0 ? "length" : "num";
            editor.putInt(name, numArgv[i]);
        }
        editor.apply();
        return true;
    }

    /**
     * Generates passwords using the preferences set by
     * {@link #setPrefs(Context, ArrayList, int...)}.
     *
     * @param ctx context from which to retrieve SharedPreferences from
     *            preferences file 'PasswordGenerator'
     * @return list of generated passwords
     */
    public static ArrayList<String> generate(Context ctx) {
        SharedPreferences prefs
                = ctx.getSharedPreferences("PasswordGenerator", Context.MODE_PRIVATE);

        boolean phonemes = true;
        int pwgenFlags = DIGITS | UPPERS;

        for (char option : pwOptions.toCharArray()) {
            if (prefs.getBoolean(String.valueOf(option), false)) {
                switch(option) {
                    case '0':
                        pwgenFlags &= ~DIGITS;
                        break;
                    case 'A':
                        pwgenFlags &= ~UPPERS;
                        break;
                    case 'B':
                        pwgenFlags |= AMBIGUOUS;
                        break;
                    case 's':
                        phonemes = false;
                        // pwgenFlags = DIGITS | UPPERS;
                        break;
                    case 'y':
                        pwgenFlags |= SYMBOLS;
                        break;
                    case 'v':
                        phonemes = false;
                        pwgenFlags |= NO_VOWELS; // | DIGITS | UPPERS;
                        break;
                }
            }
        }

        int length = prefs.getInt("length", 8);
        if (length < 5) {
            phonemes = false;
        }
        if (length <= 2) {
            pwgenFlags &= ~UPPERS;
        }
        if (length <= 1) {
            pwgenFlags &= ~DIGITS;
        }

        ArrayList<String> passwords = new ArrayList<>();
        int num = prefs.getInt("num", 1);
        for (int i = 0; i < num; i++) {
            if (phonemes) {
                passwords.add(Phonemes.phonemes(length, pwgenFlags));
            } else {
                passwords.add(RandomPasswordGenerator.rand(length, pwgenFlags));
            }
        }
        return passwords;
    }

}

File: app/src/main/java/com/zeapo/pwdstore/pwgen/PasswordGenerator.kt
/*
 * Copyright © 2014-2020 The Android Password Store Authors. All Rights Reserved.
 * SPDX-License-Identifier: GPL-3.0-only
 */
package com.zeapo.pwdstore.pwgen

import android.content.Context
import androidx.core.content.edit
import com.zeapo.pwdstore.R
import com.zeapo.pwdstore.utils.PreferenceKeys
import com.zeapo.pwdstore.utils.clearFlag
import com.zeapo.pwdstore.utils.hasFlag

enum class PasswordOption(val key: String) {
    NoDigits("0"),
    NoUppercaseLetters("A"),
    NoAmbiguousCharacters("B"),
    FullyRandom("s"),
    AtLeastOneSymbol("y"),
    NoLowercaseLetters("L")
}

object PasswordGenerator {
    const val DEFAULT_LENGTH = 16

    const val DIGITS = 0x0001
    const val UPPERS = 0x0002
    const val SYMBOLS = 0x0004
    const val NO_AMBIGUOUS = 0x0008
    const val LOWERS = 0x0020

    const val DIGITS_STR = "0123456789"
    const val UPPERS_STR = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    const val LOWERS_STR = "abcdefghijklmnopqrstuvwxyz"
    const val SYMBOLS_STR = "!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~"
    const val AMBIGUOUS_STR = "B8G6I1l0OQDS5Z2"

    /**
     * Enables the [PasswordOption]s in [options] and sets [targetLength] as the length for
     * generated passwords.
     */
    fun setPrefs(ctx: Context, options: List<PasswordOption>, targetLength: Int): Boolean {
        ctx.getSharedPreferences("PasswordGenerator", Context.MODE_PRIVATE).edit {
            for (possibleOption in PasswordOption.values())
                putBoolean(possibleOption.key, possibleOption in options)
            putInt("length", targetLength)
        }
        return true
    }

    fun isValidPassword(password: String, pwFlags: Int): Boolean {
        if (pwFlags hasFlag DIGITS && password.none { it in DIGITS_STR })
            return false
        if (pwFlags hasFlag UPPERS && password.none { it in UPPERS_STR })
            return false
        if (pwFlags hasFlag LOWERS && password.none { it in LOWERS_STR })
            return false
        if (pwFlags hasFlag SYMBOLS && password.none { it in SYMBOLS_STR })
            return false
        if (pwFlags hasFlag NO_AMBIGUOUS && password.any { it in AMBIGUOUS_STR })
            return false
        return true
    }

    /**
     * Generates a password using the preferences set by [setPrefs].
     */
    @Throws(PasswordGeneratorException::class)
    fun generate(ctx: Context): String {
        val prefs = ctx.getSharedPreferences("PasswordGenerator", Context.MODE_PRIVATE)
        var numCharacterCategories = 0

        var phonemes = true
        var pwgenFlags = DIGITS or UPPERS or LOWERS

        for (option in PasswordOption.values()) {
            if (prefs.getBoolean(option.key, false)) {
                when (option) {
                    PasswordOption.NoDigits -> pwgenFlags = pwgenFlags.clearFlag(DIGITS)
                    PasswordOption.NoUppercaseLetters -> pwgenFlags = pwgenFlags.clearFlag(UPPERS)
                    PasswordOption.NoLowercaseLetters -> pwgenFlags = pwgenFlags.clearFlag(LOWERS)
                    PasswordOption.NoAmbiguousCharacters -> pwgenFlags = pwgenFlags or NO_AMBIGUOUS
                    PasswordOption.FullyRandom -> phonemes = false
                    PasswordOption.AtLeastOneSymbol -> {
                        numCharacterCategories++
                        pwgenFlags = pwgenFlags or SYMBOLS
                    }
                }
            } else {
                // The No* options are false, so the respective character category will be included.
                when (option) {
                    PasswordOption.NoDigits,
                    PasswordOption.NoUppercaseLetters,
                    PasswordOption.NoLowercaseLetters -> {
                        numCharacterCategories++
                    }
                    PasswordOption.NoAmbiguousCharacters,
                    PasswordOption.FullyRandom,
                        // Since AtLeastOneSymbol is not negated, it is counted in the if branch.
                    PasswordOption.AtLeastOneSymbol -> {
                    }
                }
            }
        }

        val length = prefs.getInt(PreferenceKeys.LENGTH, DEFAULT_LENGTH)
        if (pwgenFlags.clearFlag(NO_AMBIGUOUS) == 0) {
            throw PasswordGeneratorException(ctx.resources.getString(R.string.pwgen_no_chars_error))
        }
        if (length < numCharacterCategories) {
            throw PasswordGeneratorException(ctx.resources.getString(R.string.pwgen_length_too_short_error))
        }
        if (!(pwgenFlags hasFlag UPPERS) && !(pwgenFlags hasFlag LOWERS)) {
            phonemes = false
            pwgenFlags = pwgenFlags.clearFlag(NO_AMBIGUOUS)
        }
        // Experiments show that phonemes may require more than 1000 iterations to generate a valid
        // password if the length is not at least 6.
        if (length < 6) {
            phonemes = false
        }

        var password: String?
        var iterations = 0
        do {
            if (iterations++ > 1000)
                throw PasswordGeneratorException(ctx.resources.getString(R.string.pwgen_max_iterations_exceeded))
            password = if (phonemes) {
                RandomPhonemesGenerator.generate(length, pwgenFlags)
            } else {
                RandomPasswordGenerator.generate(length, pwgenFlags)
            }
        } while (password == null)
        return password
    }

    class PasswordGeneratorException(string: String) : Exception(string)
}
